worldengine/draw.py

```diff
@@ -201,34 +268,424 @@ def draw_riversmap(world, target):
+                            all_r.append(r)
+                            all_g.append(g)
+                            all_b.append(b)
+
+                # Making sure there is at least one valid tile to be smoothed before we attempt to average the values
+                if all_r:
+                    avg_r = int(sum(all_r) / len(all_r))
+                    avg_g = int(sum(all_g) / len(all_g))
+                    avg_b = int(sum(all_b) / len(all_b))
+
+                    ## Setting color of the pixel again - this will be once more modified by the shading algorithm
+                    target.set_pixel(x, y, (avg_r, avg_g, avg_b, 255))
+
+    
+    # How many tiles to average together when comparing this tile's elevation to the previous tiles.
+    shade_size = 5
+    # How much to multiply the difference in elevation between this tile and the previous tile
+    # Higher will result in starker contrast between high and low areas.
+    difference_multiplier = 9
+
+    # "Shade" the map by sending beams of light west to east, and increasing or decreasing value of pixel based on elevation difference
+    for y in xrange(shade_size-1, world.height-shade_size-1):
+        for x in xrange(shade_size-1, world.width-shade_size-1):
+            if world.is_land((x, y)):
+                r, g, b, a = target.get_pixel(x, y)
+                
+                # Build up list of elevations in the previous n tiles, where n is the shadow size.
+                # This goes left to right, so it would be the previous tiles on the same y level  
+                prev_elevs = [ world.elevation['data'][y-n, x-n] for n in xrange(1, shade_size+1)]
+
+                # Take the average of the height of the previous n tiles
+                avg_prev_elev = int( sum(prev_elevs) / len(prev_elevs) )
+
+                # Find the difference between this tile's elevation, and the average of the previous elevations
+                difference = int(world.elevation['data'][y, x] - avg_prev_elev)
+
+                # Amplify the difference
+                difference = difference * difference_multiplier
+
+                # The amplified difference is now translated into the rgb of the tile.
+                # This adds light to tiles higher that the previous average, and shadow
+                # to tiles lower than the previous average
+                r += difference
+                g += difference
+                b += difference
+
+                # Set the final color for this pixel
+                target.set_pixel(x, y, (r, g, b, 255))
+
+
+def draw_satellite(world, target):
+    # Not sure what this first part is doing - copied from draw_greyscale_heightmap 
+    # I believe it is normalizing the heightmap values between 1 and 255?
+    # This function will use those values later, so this loop is kept in
+    min_elev_sea = None
+    max_elev_sea = None
+    min_elev_land = None
+    max_elev_land = None
```

Comment:
Again, recopy the code. ;) (It is just looking for min- and max-values in the elevation-array.)